docs/dcp2_system_design.rst

```diff
@@ -935,6 +1028,414 @@ import tool with be requested manually in #ingest-to-tdr-shared Slack channel
+   updating *only* the content type of a file see `Update the content type of
+   a data file`_.
+
+3. Update the entity describing the data file, as specified in Update an
+   entity. Do so even if the entity document itself did not change. This is
+   necessary so that the ``entity_type``, ``entity_id`` and ``version`` fields
+   in the name of the descriptor object match those in the name of the entity
+   object.
+
+Renaming a data file may result in an unreferenced TDR Firestore entry that
+should be garbage collected after making sure that 1) it is not the most
+recent version and 2) that no snapshots refer to it. Since the badly named
+``file_name`` descriptor property is actually a relative path, we define
+*renaming* of a data file to be a change to any component of that path. This
+includes moving a file to a different directory, and could result in empty
+Firestore directories that must be garbage collected as well.
+
+.. [#]
+   It would be difficult to specify a similar constraint for the ``s3_etag``
+   property of the file descriptor, simply because a multi-part S3 ETAG hash
+   is ambiguous: there can more than one such hash for any given input.
+
+.. [#]
+   This is assuming that the probability of two files of different sizes
+   having the same hash is infinitesimally small, except for ``crc32c``, which
+   has a significantly higher probability of collisions
+
+
+Update the content type of a data file
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Write the updated descriptor to an object in the ``descriptors`` directory
+   of the staging area. The name of the descriptor object must have the same
+   ``entity_type`` and ``entity_id`` as the original and a strictly higher
+   ``version``. The ``file_id`` and ``file_version`` properties of the updated
+   descriptor must be the same as in the previous version. Note that the
+   ``file_id``property of the descriptor document is independent from the column
+   of the same name in TDR ``…_file`` tables and that this section only applies
+   to the former.
+
+2. Update the entity describing the data file, as specified in `Update an
+   entity`_. Do so even if the entity document itself did not change. This is
+   necessary so that the ``entity_type``, ``entity_id`` and ``version`` fields
+   in the name of the descriptor object match those in the name of the
+   object containing the entity that describes the data file.
+
+
+Update a subgraph
+~~~~~~~~~~~~~~~~~
+
+Write the updated subgraph to an object in the ``links`` directory of the
+staging area. The name of the subgraph object must have the same ``links_id``
+and ``project_id`` as the original and a strictly higher ``version``.
+
+Note that it is uncommon for a subgraph to change without entities being added
+or removed. The scenarios during which it occurs naturally include the
+reordering of entities in the subgraph or migrating the subgraph document to a
+newer schema version.
```

Comment:
This question might be out of the scope of this spec. But, I'm just thinking, I haven't seen the analysis subgraphs yet, but will there ever be a chance that the subgraph updates by Ingest will conflict with the analysis subgraphs? MIght be something that we can deal with later when it actually happens.

Reply:
Interesting point. Here's what would happen with the current implementation: If the primary subgraph was altered to, say, remove an output sequence file, the Azul indexer would fail to index the stitched-on analysis subgraph and the analysis outputs (matrices etc) would disappear from the index and the Data Browser display. IOW, the snapshot must be considered broken. My team would notice the failure and escalate them in the dcp2 channel. Lantern (cc: @kbergin & @jessicaway) would then have to follow-up with updated analysis subgraphs. We would then cut a new snapshot and release that instead of the broken one. It's inelegant/inefficient to have to go through two snapshots but I see no other way since Lantern's analysis is initiated with a hand-off from the Data Browser which only works if there is a snapshot to fill the index with, even if the snapshot lacks referential integrity. I almost want to include this as a non-normative section in the spec.

Reply:
Would we still require this workaround if analysis were automated and required a snapshot but not indexing/handoff from the browser?

Reply:
Not 100% sure what your're referring to with "this workaround" but if it's the fact that we have to cut two snapshots, then no, skipping Azul and DB hand-off would still require two snapshots. The only thing that would eliminate the need to cut two snapshots is if analysis used the TDR dataset as input instead of a snapshot or Azul's index. 

I now remember that I had already considered an alternative: to tile the snapshots rather than overlapping them, so the right instead of the left option:

```
┌──────────────────┐     ┌──────────────┐
│   Snapshot 2     │     │  Snapshot 1  │
│                  │     │              │
│ ┌──────────────┐ │     │ ┌──────────┐ │
│ │  Snapshot 1  │ │     │ │ Primary  │ │
│ │              │ │     │ │ subgraph │ │
│ │ ┌──────────┐ │ │     │ └──────────┘ │
│ │ │ Primary  │ │ │     │              │
│ │ │ subgraph │ │ │     └──────────────┘
│ │ └──────────┘ │ │
│ │              │ │
│ └──────────────┘ │     ┌───────────────┐
│                  │     │  Snapshot 2   │
│                  │     │               │
│   ┌───────────┐  │     │ ┌───────────┐ │
│   │ Secondary │  │     │ │ Secondary │ │
│   │ subgraph  │  │     │ │ subgraph  │ │
│   └───────────┘  │     │ └───────────┘ │
│                  │     │               │
└──────────────────┘     └───────────────┘
```

Still two snapshots but no redundant work.

Reply:
> If the primary subgraph was altered to, say, remove an output sequence file ...

This is just one case. 

Another example and this might rarely happen, but in case there is a specimen -> cell suspension change (or any changes in the upper graph).

I wonder how analysis subgraphs are created, do they only append more entities in the primary subgraphs? or do they copy the links.json from ingest and update it? In that case when a specimen -> cell suspension change, the analysis subgraph will be outdated. Although, that actually may require reanalysis and new analysis metadata, files and subgraph. In which case, I believe it'll also follow that same process you have described for them to be updated.

Reply:
>> If the primary subgraph was altered to, say, remove an output sequence file ...

> This is just one case. 

But it is the only relevant one. What happens in the interior of the subgraph doesn't matter. Subgraph stitching only relies on the referential integrity of the final outputs in a primary subgraph.

> Another example and this might rarely happen, but in case there is a specimen -> cell suspension change (or any changes in the upper graph).

It doesn't matter unless it affects the number or identity of the final outputs. And if it does, it's the same as your first example. It may be different for Ingest but in the context of this spec, that would be an unnecessary distinction.

> I wonder how analysis subgraphs are created, do they create new links.json files and add new linking? or do they copy the links.json from ingest and update it?

It's the former and the need for subgraph stitching is a direct consequence of that. This is part of the existing spec, almost from day one, if you catch my drift.

Reply:
I wasn't closely involved initially and I've never seen the pipeline's links.json files, so I didn't know. It actually makes more sense now why you called it _subgraph stitching_. Anyway, thanks for confirming. This is helpful.

Reply:
I'd recommend reading the entire spec. It's very dry matter but there aren't really any parts that aren't at least tangentially relevant to your team.

Reply:
Addressed